### PR TITLE
#BE-4783 Add profile creation settings

### PR DESCRIPTION
**Pull request by hiilhan**

### CODE REVIEW

**Review by burakberkkeskin — APPROVED** [review]
Tested on dev environment:
- The profile created with static username password successfully. Shared with email and downloaded successfully.
- The profile created with non-auth successfully. Shared with email and downloaded successfully.

File: README.md [hunk 1]
```diff
@@ -80,28 +67,43 @@ fastlane add_plugin appcircle_testing_distribution
 ```yml
   appcircle_testing_distribution(
     personalAPIToken: "$(AC_PERSONAL_API_TOKEN)",
+    subOrganizationName: "${AC_SUB_ORGANIZATION_NAME}",
```
Comment (csonuryilmaz) [inline README.md]: `{` vs. `(` Which one should be?

File: README.md [hunk 2]
```diff
@@ -80,28 +67,43 @@ fastlane add_plugin appcircle_testing_distribution
...
     profileName: "$(AC_PROFILE_NAME)",
-    createProfileIfNotExists: Boolean,
+    createProfileIfNotExists: "${AC_CREATE_PROFILE_IF_NOT_EXISTS}",
+    profileCreationSettings: {
+      authType: "${AC_PROFILE_AUTH_TYPE}",
+      username: "${AC_PROFILE_USERNAME}",
+      password: "${AC_PROFILE_PASSWORD}",
+      testingGroupNames: "AC_PROFILE_TESTING_GROUP_NAMES"
```
Comment (csonuryilmaz) [inline README.md]: Fix

File: README.md [hunk 3]
```diff
@@ -80,28 +67,43 @@ fastlane add_plugin appcircle_testing_distribution
...
+- `profileCreationSettings` (optional): If `createProfileIfNotExists` is `true` and a new profile being created, the profile will be configured with these settings.
+   - `authType`: Authentication type of the profile. `none`: None, `static`: Static Username and Password, `ldap`: LDAP Login, `sso`: SSO Login. If this variable is not defined, the profile's authentication type will also be undefined and login will not be possible.
+   - `username`: The username for the profile if authentication type set to `static` (Static Username and Password).
+   - `password`: The password for the profile if authentication type set to `static` (Static Username and Password).
+   - `testingGroupNames`: Uploaded versions will be automatically shared with these testing groups. Example format: `group1, group2, group3`.
 - `appPath`: Indicates the file path to the application that will be uploaded to Appcircle Testing Distribution Profile.
 - `message`: Your message to testers, ensuring they receive important updates and information regarding the application.
 
-### Leveraging Environment Variables
+## Docs
```
Comment (csonuryilmaz) [inline README.md]:
```suggestion
## Further Details
```

**Review by csonuryilmaz — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -79,29 +66,44 @@ fastlane add_plugin appcircle_testing_distribution
 
 ```yml
```
Comment (csonuryilmaz) [inline README.md]: It should be `ruby` instead of yaml.

File: lib/fastlane/plugin/appcircle_testing_distribution/helper/TDUploadService.rb [hunk 5]
```diff
@@ -10,7 +11,7 @@ def self.upload_artifact(token:, message:, app:, dist_profile_id:)
     url = "https://api.appcircle.io/distribution/v2/profiles/#{dist_profile_id}/app-versions"
     headers = {
       Authorization: "Bearer #{token}",
-      content_type: :multipart # multipart/form-data
+      # content_type: :multipart or multipart/form-data # Causing a type cast warning
```
Comment (csonuryilmaz) [inline lib/fastlane/plugin/appcircle_testing_distribution/helper/TDUploadService.rb]: ❓ Are we sure about this change? Does it affect anything?

Content type might be an important argument in request headers.

File: lib/fastlane/plugin/appcircle_testing_distribution/helper/TDUploadService.rb [hunk 6]
```diff
@@ -2,6 +2,7 @@
 require 'uri'
 require 'json'
 require 'rest-client'
+require_relative '../helper/appcircle_testing_distribution_helper'
```
Comment (csonuryilmaz) [inline lib/fastlane/plugin/appcircle_testing_distribution/helper/TDUploadService.rb]: AFAIK, we do not refer to any function from this helper ruby. So, it seems to be unnecessary to "require" the file here.

File: README.md [hunk 7]
```diff
@@ -79,29 +66,44 @@ fastlane add_plugin appcircle_testing_distribution
...
-
-**Ensure that this action is added after build steps have been completed.**
-
-**If multiple workflows start simultaneously, the order in which versions are shared in the Testing Distribution is determined by the execution order of the publish step. The version that completes its build and triggers the publish plugin first will be shared first, followed by the others in sequence.**
-
-Efficiently distribute test binaries or beta versions using Appcircle, featuring seamless IPA and APK distribution capabilities. Streamline your testing process with our versatile tool designed to optimize your distribution workflow. If you need support or more information, please [contact us](https://appcircle.io/contact?utm_source=fastlane&utm_medium=plugin&utm_campaign=testing_distribution)
+- [Setting Up Appcircle Testing Distribution Plugin](https://docs.appcircle.io/marketplace/fastlane/testing-distribution)
+   - [Discover Action](https://docs.appcircle.io/marketplace/fastlane/testing-distribution#discover-action)
```
Comment (csonuryilmaz) [inline README.md]: Markdown indentation should be 2 spaces instead of 3 spaces for unordered list items. (markdown lint)

File: README.md [hunk 8]
```diff
@@ -79,29 +66,44 @@ fastlane add_plugin appcircle_testing_distribution
...
 - `profileName`: Specifies the profile that will be used for uploading the app.
-- `createProfileIfNotExists`: Ensures that a user profile is automatically created if it does not already exist; if the profile name already exists, the app will be uploaded to that existing profile instead.
-- `appPath`: Indicates the file path to the application that will be uploaded to Appcircle Testing Distribution Profile.
+- `createProfileIfNotExists` (optional): Ensures that a testing distribution profile is automatically created if it does not already exist; if the profile name already exists, the app will be uploaded to that existing profile instead.
+- `profileCreationSettings` (optional): If `createProfileIfNotExists` is `true` and a new profile being created, the profile will be configured with these settings.
+   - `authType`: Authentication type of the profile. `none`: None, `static`: Static Username and Password, `ldap`: LDAP Login, `sso`: SSO Login.
```
Comment (csonuryilmaz) [inline README.md]: Markdown indentation should be 2 spaces instead of 3 spaces for unordered list items. (markdown lint)

**Review by csonuryilmaz — CHANGES_REQUESTED** [review]
@hiilhan I have added some threads to be resolved and one question to discuss.

In addition to those, please also fix the cases below.

1️⃣ Remove obsolete images from repository.
- `images/distribution-start.png`
- `images/PAT.png`

AFAIK, we are using the CDN URL in the README. Also, they might be stale.

2️⃣ The license should be owned by Appcircle instead of an individual contributor. You can check other samples from workflow step license files.

Screenshot 2024-12-18 at 03 25 01

3️⃣ Update the `version.rb` file to our target version: `0.3.2`. So, our PR will be ready for release when merged.

File: README.md (hunk 4 above)
Comment (hiilhan) [reply README.md]: 1. Done
2. Done
3. Updated to 0.3.2-dev

File: lib/fastlane/plugin/appcircle_testing_distribution/helper/TDUploadService.rb (hunk 5 above)
Comment (hiilhan) [reply lib/fastlane/plugin/appcircle_testing_distribution/helper/TDUploadService.rb]: Didn't affect the tests, I will check again.

File: fastlane/Fastfile [hunk 9]
```diff
@@ -4,9 +4,16 @@ Dotenv.load
 lane :test do
   appcircle_testing_distribution(
     personalAPIToken: ENV["AC_ACCESS_TOKEN"],
```
Comment (csonuryilmaz) [inline fastlane/Fastfile]: Change to `AC_PERSONAL_API_TOKEN`

File: README.md [hunk 10]
```diff
@@ -65,43 +52,59 @@ To generate a Personal API Token, follow these steps:
...
 
-```yml
+```ruby
   appcircle_testing_distribution(
-    personalAPIToken: "$(AC_PERSONAL_API_TOKEN)",
-    profileName: "$(AC_PROFILE_NAME)",
-    createProfileIfNotExists: Boolean,
-    appPath: "$(AC_APP_PATH)",
-    message: "$(AC_MESSAGE)",
+    personalAPIToken: ENV["AC_ACCESS_TOKEN"],
```
Comment (csonuryilmaz) [inline README.md]: Change to `AC_PERSONAL_API_TOKEN`

File: lib/fastlane/plugin/appcircle_testing_distribution/helper/TDUploadService.rb (hunk 5 above)
Comment (hiilhan) [reply lib/fastlane/plugin/appcircle_testing_distribution/helper/TDUploadService.rb]: > Auto-detects that you have a File value sends it as multipart
> (from https://github.com/rest-client/rest-client#multipart)

Current payload has a File value but added `multipart: true`, to make sure. Tested the result, request didn't change.

Comment (csonuryilmaz) [reply lib/fastlane/plugin/appcircle_testing_distribution/helper/TDUploadService.rb]: OK. Thanks for the research and clarification.

**Review by csonuryilmaz — APPROVED** [review]

**Pull request merged**